core/executor/src/wasm_runtime.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements a cache for pre-created Wasm runtime module instances.
+
+use crate::error::{Error, WasmError};
+use crate::wasmi_execution;
+use log::{trace, warn};
+use codec::Decode;
+use primitives::{storage::well_known_keys, Blake2Hasher, traits::Externalities};
+use runtime_version::RuntimeVersion;
+use std::{collections::hash_map::{Entry, HashMap}};
+
+/// The Wasm Substrate runtime.
+pub trait WasmRuntime {
+	/// Attempt to set the number of heap pages available during execution.
+	///
+	/// Returns false if the heap pages cannot be changed to the given value. Always returns true
```

Comment:
```suggestion
	/// Returns `false` if the heap pages cannot be changed to the given value. Always returns `true`
```

Reply:
I also can not really follow. "Always returns true if the heap pages is set to the value already configured" and it returns `false` when it can not change the value. What does it return when it could change the value?

Reply:
Ahh, reading the implementation, I think that I understand the idea behind the function. I think it should be renamed. Maybe something like "heap_pages_equals" or so? As you just want to know that the heap pages are at the same value you expect.

Reply:
Well, the reason for this is that some runtimes may be able to update the heap pages and some may not. Specifically, the heap pages is fixed with wasmi and updateable with wasmtime. But you could also consider that maybe it's possible for some engine to grow the heap pages but not shrink them.

So the point is that this attempts to set the heap pages to the new value and returns false if the update could not be performed. And it returns true if the heap pages would not change as a result of the update (this is guaranteed for all runtime implementations).

Reply:
Then something like `update_heap_pages`? Your description is much better than the current documentation of the function + name.

```
/// Try to update the heap pages.
///
/// Returns `true` if the update could be applied successfully or `false` otherwise.
fn update_heap_pages(heap_pages: usize);
```

Maybe something like this?